File: sanguo/tags/for_vr/client/gateclient.cpp

#include "gateclient.hpp"
#include "state.hxx"

#include "connection.h"

namespace GNET
{

GateClient GateClient::instance;

const Protocol::Manager::Session::State* GateClient::GetInitState() const
{
	return &state_GateClient;
}

void GateClient::OnAddSession(Session::ID sid)
{
	fprintf(stderr, "GateClient::OnAddSession, this->sid=%d, sid=%d\n", this->sid, sid);

	Thread::Mutex::Scoped l(locker_state);

	if(conn_state)
	{
		Close(sid);
		return;
	}
	if(_discard_sid_le_than>0 && sid<=_discard_sid_le_than)
	{
		Close(sid);
		return;
	}
	conn_state = true;
	this->sid = sid;
}

void GateClient::OnDelSession(Session::ID sid)
{
	fprintf(stderr, "GateClient::OnDelSession, this->sid=%d, sid=%d\n", this->sid, sid);

	Thread::Mutex::Scoped l(locker_state);

	if(!conn_state || this->sid!=sid) return;
	conn_state = false;
}

void GateClient::OnAbortSession(const SockAddr &sa)
{
	//Thread::Mutex::Scoped l(locker_state);
	//conn_state = false;
}

void GateClient::OnCheckAddress(SockAddr &sa) const
{
}

bool GateClient::LoadSessionConfig(GNET::NetSession::Config &cnf)
{
	if(!GNET::Protocol::Manager::LoadSessionConfig(cnf)) return false;
	cnf.address = Connection::GetInstance().GetGateIP();
	cnf.port = Connection::GetInstance().GetGatePort();
	return true;
}

void GateClient::CloseCur()
{
	if (conn_state)
	{
		conn_state = false;
		Close(sid);
	}
}

int GateClient::GetCurSID() const
{
	if(!conn_state) return 0;
	return sid;
}

};

File: sanguo/tags/banshu_1609/csharp/client/csharp/Manager/GateClient.cs
using System;
using GNET.IO;
using GNET.Common;

namespace GNET
{
	public class GateClient : Manager
	{
		private Session session = null;

		private static readonly GateClient instance = new GateClient();
		public static GateClient GetInstance()
		{
			return instance;
		}

		public Session GetSession() { return session; }
		protected internal override void OnAddSession (Session session)
		{
			Console.WriteLine("GateClient::OnAddSession");
			this.session = session;
		}

		protected internal override void OnDelSession (Session session)
		{
			if(this.session != null)
				Console.WriteLine("GateClient::OnDelSession " + this.session.getPeerAddress ().ToString ());
			this.session = null;
			//reconnect
		}

		protected internal override void OnAbortSession (Session session)
		{
			if(this.session != null)
				Console.WriteLine("GateClient::OnAbortSession " + this.session.getPeerAddress ().ToString ());
			this.session = null;
			//reconnect
		}

		public override String Identification()
		{
			return "GateClient";
		}

		public bool SendProtocol(Protocol protocol) { return (session != null) && Send(session, protocol); }
		public override void OnRecvProtocol(Session session, Protocol protocol)
		{
			//TODO
		}

	}
}
